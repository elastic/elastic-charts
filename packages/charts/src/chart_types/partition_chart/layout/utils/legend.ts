/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { CategoryKey } from '../../../../common/category';
import { map } from '../../../../common/iterables';
import { LegendItem } from '../../../../common/legend';
import { LegendPositionConfig } from '../../../../specs/settings';
import { isHierarchicalLegend } from '../../../../utils/legend';
import { Layer } from '../../specs';
import { PartitionLayout } from '../types/config_types';
import { QuadViewModel } from '../types/viewmodel_types';

function makeKey(...keyParts: CategoryKey[]): string {
  return keyParts.join('---');
}

function compareTreePaths(
  { index: oiA, innerIndex: iiA, path: a }: QuadViewModel,
  { index: oiB, innerIndex: iiB, path: b }: QuadViewModel,
): number {
  if (oiA !== oiB) return oiA - oiB;
  if (iiA !== iiB) return iiA - iiB;
  for (let i = 0; i < Math.min(a.length, b.length); i++) {
    const diff = a[i].index - b[i].index;
    if (diff) {
      return diff;
    }
  }
  return a.length - b.length; // if one path is fully contained in the other, then parent (shorter) goes first
}

/** @internal */
export function getLegendItems(
  id: string,
  layers: Layer[],
  flatLegend: boolean | undefined,
  legendMaxDepth: number,
  legendPosition: LegendPositionConfig,
  quadViewModel: QuadViewModel[],
  partitionLayout: PartitionLayout | undefined,
): LegendItem[] {
  const uniqueNames = new Set(map(quadViewModel, ({ dataName, fillColor }) => makeKey(dataName, fillColor)));
  const useHierarchicalLegend = isHierarchicalLegend(flatLegend, legendPosition);

  const formattedLabel = ({ dataName, depth }: QuadViewModel) => {
    const formatter = layers[depth - 1]?.nodeLabel;
    return formatter ? formatter(dataName) : dataName;
  };

  function compareNames(aItem: QuadViewModel, bItem: QuadViewModel): number {
    const a = formattedLabel(aItem);
    const b = formattedLabel(bItem);
    return a < b ? -1 : a > b ? 1 : 0;
  }

  function descendingValues(aItem: QuadViewModel, bItem: QuadViewModel): number {
    return aItem.depth - bItem.depth || bItem.value - aItem.value;
  }

  const excluded: Set<string> = new Set();
  const items = quadViewModel.filter(({ depth, dataName, fillColor }) => {
    if (legendMaxDepth !== null && depth > legendMaxDepth) {
      return false;
    }
    if (!useHierarchicalLegend) {
      const key = makeKey(dataName, fillColor);
      if (uniqueNames.has(key) && excluded.has(key)) {
        return false;
      }
      excluded.add(key);
    }
    return true;
  });

  items.sort(
    partitionLayout === PartitionLayout.waffle // waffle has inherent top to bottom descending order
      ? descendingValues
      : flatLegend
      ? compareNames
      : compareTreePaths,
  );

  return items.map<LegendItem>((item) => {
    const { dataName, fillColor, depth, path } = item;
    return {
      color: fillColor,
      label: formattedLabel(item),
      childId: dataName,
      depth: useHierarchicalLegend ? depth - 1 : 0,
      path,
      seriesIdentifiers: [{ key: dataName, specId: id }],
      keys: [],
    };
  });
}
