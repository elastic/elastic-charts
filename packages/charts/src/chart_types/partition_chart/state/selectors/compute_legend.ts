/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { LegendItem } from '../../../../common/legend';
import { createCustomCachedSelector } from '../../../../state/create_selector';
import { getLegendConfigSelector } from '../../../../state/selectors/get_legend_config_selector';
import { getLegendItems } from '../../layout/utils/legend';
import { partitionMultiGeometries } from './geometries';
import { getPartitionSpecs } from './get_partition_specs';

/** @internal */
export const computeLegendSelector = createCustomCachedSelector(
  [getPartitionSpecs, getLegendConfigSelector, partitionMultiGeometries],
  (specs, { flatLegend, legendMaxDepth, legendPosition }, geometries): LegendItem[] =>
    specs.flatMap((partitionSpec, i) => {
      const quadViewModel = geometries.filter((g) => g.index === i).flatMap((g) => g.quadViewModel);
      return getLegendItems(
        partitionSpec.id,
        partitionSpec.layers,
        flatLegend,
        legendMaxDepth,
        legendPosition,
        quadViewModel,
      );
    }),
);
