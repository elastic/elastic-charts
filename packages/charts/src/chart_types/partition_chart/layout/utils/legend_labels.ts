/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { LegendItemLabel } from '../../../../state/selectors/get_legend_items_labels';
import { ValueFormatter } from '../../../../utils/common';
import { Layer } from '../../specs';
import {
  CHILDREN_KEY,
  HIERARCHY_ROOT_KEY,
  HierarchyOfArrays,
  PATH_KEY,
  ArrayNode,
  NULL_SMALL_MULTIPLES_KEY,
  AGGREGATE_KEY,
} from './group_by_rollup';

/** @internal */
export function getLegendLabelsAndValue(
  layers: Layer[],
  tree: HierarchyOfArrays,
  legendMaxDepth: number,
  // if used the resulting label will be concatenated with the formatted value, use () => '' to avoid that
  valueFormatter: ValueFormatter,
): LegendItemLabel[] {
  return flatSlicesNames(layers, 0, tree, valueFormatter).filter(({ depth }) => depth <= legendMaxDepth);
}

/** @internal */
function getArrayNodeKey(arrayNode: ArrayNode): string {
  return arrayNode[PATH_KEY].reduce<string>((acc, { value, index }) => {
    if (value === HIERARCHY_ROOT_KEY || value === NULL_SMALL_MULTIPLES_KEY) return acc;
    return `${acc}(${index}):${value}__`;
  }, '__');
}

function flatSlicesNames(
  layers: Layer[],
  depth: number,
  tree: HierarchyOfArrays,
  valueFormatter: ValueFormatter,
  keys: Map<string, string> = new Map(),
  depths: Map<string, number> = new Map(),
): LegendItemLabel[] {
  if (tree.length === 0) {
    return [];
  }

  for (const [key, arrayNode] of tree) {
    // format the key with the layer formatter
    const layer = layers[depth - 1];
    const formatter = layer?.nodeLabel;
    const formattedKey = formatter ? formatter(key) : `${key}`;
    // preventing errors from external formatters
    if (formattedKey && formattedKey !== HIERARCHY_ROOT_KEY) {
      // Node key must be unique for each node in the tree
      const nodeKey = getArrayNodeKey(arrayNode);
      // save only the max depth, so we can compute the the max extension of the legend
      depths.set(nodeKey, depth);

      const formattedValue = valueFormatter(arrayNode[AGGREGATE_KEY]);
      keys.set(nodeKey, `${formattedKey}${formattedValue}`);
    }

    const children = arrayNode[CHILDREN_KEY];
    flatSlicesNames(layers, depth + 1, children, valueFormatter, keys, depths);
  }

  return [...depths.keys()].map((key) => ({
    label: keys.get(key) ?? '',
    depth: depths.get(key) ?? 0,
  }));
}
