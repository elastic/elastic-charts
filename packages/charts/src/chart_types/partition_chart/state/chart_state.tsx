/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { RefObject } from 'react';

import { ChartType } from '../..';
import { BackwardRef, GlobalChartState, InternalChartState } from '../../../state/chart_state';
import { getActivePointerPosition } from '../../../state/selectors/get_active_pointer_position';
import { InitStatus } from '../../../state/selectors/get_internal_is_intialized';
import { DebugState } from '../../../state/types';
import { Dimensions } from '../../../utils/dimensions';
import { render } from '../renderer/dom/layered_partition_chart';
import { computeLegendSelector } from './selectors/compute_legend';
import { getChartTypeDescriptionSelector } from './selectors/get_chart_type_description';
import { getPointerCursorSelector } from './selectors/get_cursor_pointer';
import { getDebugStateSelector } from './selectors/get_debug_state';
import { getLegendItemsExtra } from './selectors/get_legend_items_extra';
import { getLegendItemsLabels } from './selectors/get_legend_items_labels';
import { isTooltipVisibleSelector } from './selectors/is_tooltip_visible';
import { createOnElementClickCaller } from './selectors/on_element_click_caller';
import { createOnElementOutCaller } from './selectors/on_element_out_caller';
import { createOnElementOverCaller } from './selectors/on_element_over_caller';
import { getPartitionSpec } from './selectors/partition_spec';
import { getTooltipInfoSelector } from './selectors/tooltip';

/** @internal */
export class PartitionState implements InternalChartState {
  chartType = ChartType.Partition;

  onElementClickCaller: (state: GlobalChartState) => void;

  onElementOverCaller: (state: GlobalChartState) => void;

  onElementOutCaller: (state: GlobalChartState) => void;

  constructor() {
    this.onElementClickCaller = createOnElementClickCaller();
    this.onElementOverCaller = createOnElementOverCaller();
    this.onElementOutCaller = createOnElementOutCaller();
  }

  isInitialized(globalState: GlobalChartState) {
    return getPartitionSpec(globalState) !== null ? InitStatus.Initialized : InitStatus.SpecNotInitialized;
  }

  isBrushAvailable() {
    return false;
  }

  isBrushing() {
    return false;
  }

  isChartEmpty() {
    return false;
  }

  getLegendItemsLabels(globalState: GlobalChartState) {
    // order doesn't matter, but it needs to return the highest depth of the label occurrence so enough horizontal width is allocated
    // the label item strings needs to be a concatenation of the label + the extra formatted value if available.
    // this is required to compute the legend automatic width
    return getLegendItemsLabels(globalState);
  }

  getLegendItems(globalState: GlobalChartState) {
    return computeLegendSelector(globalState);
  }

  getLegendExtraValues(globalState: GlobalChartState) {
    return getLegendItemsExtra(globalState);
  }

  chartRenderer(containerRef: BackwardRef, forwardStageRef: RefObject<HTMLCanvasElement>) {
    return render(containerRef, forwardStageRef);
  }

  getPointerCursor(globalState: GlobalChartState) {
    return getPointerCursorSelector(globalState);
  }

  isTooltipVisible(globalState: GlobalChartState) {
    return {
      visible: isTooltipVisibleSelector(globalState),
      isExternal: false,
      displayOnly: false,
      isPinnable: true,
    };
  }

  getTooltipInfo(globalState: GlobalChartState) {
    return getTooltipInfoSelector(globalState);
  }

  getTooltipAnchor(state: GlobalChartState) {
    const position = getActivePointerPosition(state);
    return {
      isRotated: false,
      x: position.x,
      width: 0,
      y: position.y,
      height: 0,
    };
  }

  eventCallbacks(globalState: GlobalChartState) {
    this.onElementOverCaller(globalState);
    this.onElementOutCaller(globalState);
    this.onElementClickCaller(globalState);
  }

  // TODO
  getProjectionContainerArea(): Dimensions {
    return { width: 0, height: 0, top: 0, left: 0 };
  }

  // TODO
  getMainProjectionArea(): Dimensions {
    return { width: 0, height: 0, top: 0, left: 0 };
  }

  // TODO
  getBrushArea(): Dimensions | null {
    return null;
  }

  getDebugState(state: GlobalChartState): DebugState {
    return getDebugStateSelector(state);
  }

  getChartTypeDescription(state: GlobalChartState): string {
    return getChartTypeDescriptionSelector(state);
  }
}
