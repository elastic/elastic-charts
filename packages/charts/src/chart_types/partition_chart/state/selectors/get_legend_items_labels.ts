/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { createCustomCachedSelector } from '../../../../state/create_selector';
import { LegendItemLabel } from '../../../../state/selectors/get_legend_items_labels';
import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_spec';
import { getLegendLabelsAndValue } from '../../layout/utils/legend_labels';
import { getPartitionSpecs } from './get_partition_specs';
import { getTrees } from './tree';

/** @internal */
export const getLegendItemsLabels = createCustomCachedSelector(
  [getPartitionSpecs, getSettingsSpecSelector, getTrees],
  (specs, { legendMaxDepth, showLegend, showLegendExtra }, trees): LegendItemLabel[] =>
    specs.flatMap(({ layers, valueFormatter }) =>
      showLegend
        ? trees.flatMap(({ tree }) =>
            getLegendLabelsAndValue(layers, tree, legendMaxDepth, showLegendExtra ? valueFormatter : () => ''),
          )
        : [],
    ),
);
