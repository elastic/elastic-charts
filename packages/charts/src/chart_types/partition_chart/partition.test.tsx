/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Store } from 'redux';

import { MockGlobalSpec, MockSeriesSpec } from '../../mocks/specs';
import { MockStore } from '../../mocks/store';
import { GlobalChartState } from '../../state/chart_state';
import { LegendItemLabel } from '../../state/selectors/get_legend_items_labels';
import { HIERARCHY_ROOT_KEY } from './layout/utils/group_by_rollup';
import { computeLegendSelector } from './state/selectors/compute_legend';
import { getLegendItemsLabels } from './state/selectors/get_legend_items_labels';

// sorting is useful to ensure tests pass even if order changes (where order doesn't matter)
const ascByLabel = (a: LegendItemLabel, b: LegendItemLabel) => (a.label < b.label ? -1 : a.label > b.label ? 1 : 0);

describe('Retain hierarchy even with arbitrary names', () => {
  type TestDatum = { cat1: string; cat2: string; val: number };
  const specJSON = {
    data: [
      { cat1: 'A', cat2: 'A', val: 1 },
      { cat1: 'A', cat2: 'B', val: 1 },
      { cat1: 'B', cat2: 'A', val: 1 },
      { cat1: 'B', cat2: 'B', val: 1 },
      { cat1: 'C', cat2: 'A', val: 1 },
      { cat1: 'C', cat2: 'B', val: 1 },
    ],
    valueAccessor: (d: TestDatum) => d.val,
    layers: [
      {
        groupByRollup: (d: TestDatum) => d.cat1,
      },
      {
        groupByRollup: (d: TestDatum) => d.cat2,
      },
    ],
  };
  let store: Store<GlobalChartState>;

  beforeEach(() => {
    store = MockStore.default();
  });

  describe('getLegendItemsLabels', () => {
    // todo discuss question marks about testing this selector, and also about unification with `get_legend_items_labels.test.ts`

    it('all distinct labels are present', () => {
      MockStore.addSpecs([MockGlobalSpec.settings({ showLegend: true }), MockSeriesSpec.sunburst(specJSON)], store);
      expect(getLegendItemsLabels(store.getState()).sort(ascByLabel)).toEqual([
        { depth: 2, label: 'A' },
        { depth: 2, label: 'B' },
        { depth: 1, label: 'C' },
      ]);
    });

    it('no labels are present if showLegend is false', () => {
      MockStore.addSpecs([MockGlobalSpec.settings({ showLegend: false }), MockSeriesSpec.sunburst(specJSON)], store);
      expect(getLegendItemsLabels(store.getState())).toEqual([]);
    });

    it('no labels are present if showLegend is missing', () => {
      MockStore.addSpecs([MockSeriesSpec.sunburst(specJSON)], store);
      expect(getLegendItemsLabels(store.getState())).toEqual([]);
    });

    it('special case: one input, one label', () => {
      MockStore.addSpecs(
        [
          MockGlobalSpec.settings({ showLegend: true }),
          MockSeriesSpec.sunburst({ ...specJSON, data: [{ cat1: 'A', cat2: 'A', val: 1 }] }),
        ],
        store,
      );
      expect(getLegendItemsLabels(store.getState())).toEqual([{ depth: 2, label: 'A' }]);
    });

    it('special case: one input, two labels', () => {
      MockStore.addSpecs(
        [
          MockGlobalSpec.settings({ showLegend: true }),
          MockSeriesSpec.sunburst({ ...specJSON, data: [{ cat1: 'C', cat2: 'B', val: 1 }] }),
        ],
        store,
      );
      expect(getLegendItemsLabels(store.getState()).sort(ascByLabel)).toEqual([
        { depth: 2, label: 'B' },
        { depth: 1, label: 'C' },
      ]);
    });

    it('special case: no labels', () => {
      MockStore.addSpecs(
        [MockGlobalSpec.settings({ showLegend: true }), MockSeriesSpec.sunburst({ ...specJSON, data: [] })],
        store,
      );
      expect(getLegendItemsLabels(store.getState()).map((l) => l.label)).toEqual([]);
    });
  });

  describe('getLegendItems', () => {
    // todo discuss question marks about testing this selector, and also about unification with `get_legend_items_labels.test.ts`

    it('all distinct labels are present', () => {
      MockStore.addSpecs(
        [
          MockGlobalSpec.settings({ showLegend: true }),
          MockGlobalSpec.settings({ showLegend: true }),
          MockSeriesSpec.sunburst(specJSON),
        ],
        store,
      );
      expect(computeLegendSelector(store.getState())).toEqual([
        {
          childId: 'A',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            { index: 0, value: HIERARCHY_ROOT_KEY },
            { index: 0, value: 'A' },
          ],
          depth: 0,
          label: 'A',
          seriesIdentifiers: [{ key: 'A', specId: 'spec1' }],
          keys: [],
        },
        {
          childId: 'A',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            { index: 0, value: HIERARCHY_ROOT_KEY },
            { index: 0, value: 'A' },
            { index: 0, value: 'A' },
          ],
          depth: 1,
          label: 'A',
          seriesIdentifiers: [{ key: 'A', specId: 'spec1' }],
          keys: [],
        },
        {
          childId: 'B',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            { index: 0, value: HIERARCHY_ROOT_KEY },
            { index: 0, value: 'A' },
            { index: 1, value: 'B' },
          ],
          depth: 1,
          label: 'B',
          seriesIdentifiers: [{ key: 'B', specId: 'spec1' }],
          keys: [],
        },
        {
          childId: 'B',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            { index: 0, value: HIERARCHY_ROOT_KEY },
            { index: 1, value: 'B' },
          ],
          depth: 0,
          label: 'B',
          seriesIdentifiers: [{ key: 'B', specId: 'spec1' }],
          keys: [],
        },
        {
          childId: 'A',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            { index: 0, value: HIERARCHY_ROOT_KEY },
            { index: 1, value: 'B' },
            { index: 0, value: 'A' },
          ],
          depth: 1,
          label: 'A',
          seriesIdentifiers: [{ key: 'A', specId: 'spec1' }],
          keys: [],
        },
        {
          childId: 'B',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            { index: 0, value: HIERARCHY_ROOT_KEY },
            { index: 1, value: 'B' },
            { index: 1, value: 'B' },
          ],
          depth: 1,
          label: 'B',
          seriesIdentifiers: [{ key: 'B', specId: 'spec1' }],
          keys: [],
        },
        {
          childId: 'C',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            { index: 0, value: HIERARCHY_ROOT_KEY },
            { index: 2, value: 'C' },
          ],
          depth: 0,
          label: 'C',
          seriesIdentifiers: [{ key: 'C', specId: 'spec1' }],
          keys: [],
        },
        {
          childId: 'A',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            { index: 0, value: HIERARCHY_ROOT_KEY },
            { index: 2, value: 'C' },
            { index: 0, value: 'A' },
          ],
          depth: 1,
          label: 'A',
          seriesIdentifiers: [{ key: 'A', specId: 'spec1' }],
          keys: [],
        },
        {
          childId: 'B',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            { index: 0, value: HIERARCHY_ROOT_KEY },
            { index: 2, value: 'C' },
            { index: 1, value: 'B' },
          ],
          depth: 1,
          label: 'B',
          seriesIdentifiers: [{ key: 'B', specId: 'spec1' }],
          keys: [],
        },
      ]);
    });

    it('special case: one input, one label', () => {
      MockStore.addSpecs(
        [
          MockGlobalSpec.settings({ showLegend: true }),
          MockSeriesSpec.sunburst({ ...specJSON, data: [{ cat1: 'A', cat2: 'A', val: 1 }] }),
        ],
        store,
      );
      expect(computeLegendSelector(store.getState())).toEqual([
        {
          childId: 'A',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            {
              index: 0,
              value: HIERARCHY_ROOT_KEY,
            },
            {
              index: 0,
              value: 'A',
            },
          ],
          depth: 0,
          label: 'A',
          seriesIdentifiers: [{ key: 'A', specId: 'spec1' }],
          keys: [],
        },
        {
          childId: 'A',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            {
              index: 0,
              value: HIERARCHY_ROOT_KEY,
            },
            {
              index: 0,
              value: 'A',
            },
            {
              index: 0,
              value: 'A',
            },
          ],

          depth: 1,
          label: 'A',
          seriesIdentifiers: [{ key: 'A', specId: 'spec1' }],
          keys: [],
        },
      ]);
    });

    it('special case: one input, two labels', () => {
      MockStore.addSpecs(
        [
          MockGlobalSpec.settings({ showLegend: true }),
          MockSeriesSpec.sunburst({ ...specJSON, data: [{ cat1: 'C', cat2: 'B', val: 1 }] }),
        ],
        store,
      );
      expect(computeLegendSelector(store.getState())).toEqual([
        {
          childId: 'C',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            {
              index: 0,
              value: HIERARCHY_ROOT_KEY,
            },
            {
              index: 0,
              value: 'C',
            },
          ],
          depth: 0,
          label: 'C',
          seriesIdentifiers: [{ key: 'C', specId: 'spec1' }],
          keys: [],
        },
        {
          childId: 'B',
          color: 'rgba(128, 0, 0, 0.5)',
          path: [
            {
              index: 0,
              value: HIERARCHY_ROOT_KEY,
            },
            {
              index: 0,
              value: 'C',
            },
            {
              index: 0,
              value: 'B',
            },
          ],
          depth: 1,
          label: 'B',
          seriesIdentifiers: [{ key: 'B', specId: 'spec1' }],
          keys: [],
        },
      ]);
    });

    it('special case: no labels', () => {
      MockStore.addSpecs(
        [MockGlobalSpec.settings({ showLegend: true }), MockSeriesSpec.sunburst({ ...specJSON, data: [] })],
        store,
      );
      expect(getLegendItemsLabels(store.getState()).map((l) => l.label)).toEqual([]);
    });
  });
});
