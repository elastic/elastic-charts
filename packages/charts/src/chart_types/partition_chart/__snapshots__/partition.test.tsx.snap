// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Retain hierarchy even with arbitrary names getLegendItems all distinct labels are present 1`] = `
[
  {
    "childId": "A",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 0,
    "keys": [],
    "label": "A",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 0,
        "value": "A",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "A",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "2",
        "value": 2,
      },
    ],
  },
  {
    "childId": "A",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 1,
    "keys": [],
    "label": "A",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 0,
        "value": "A",
      },
      {
        "index": 0,
        "value": "A",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "A",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "1",
        "value": 1,
      },
    ],
  },
  {
    "childId": "B",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 1,
    "keys": [],
    "label": "B",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 0,
        "value": "A",
      },
      {
        "index": 1,
        "value": "B",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "B",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "1",
        "value": 1,
      },
    ],
  },
  {
    "childId": "B",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 0,
    "keys": [],
    "label": "B",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 1,
        "value": "B",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "B",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "2",
        "value": 2,
      },
    ],
  },
  {
    "childId": "A",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 1,
    "keys": [],
    "label": "A",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 1,
        "value": "B",
      },
      {
        "index": 0,
        "value": "A",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "A",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "1",
        "value": 1,
      },
    ],
  },
  {
    "childId": "B",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 1,
    "keys": [],
    "label": "B",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 1,
        "value": "B",
      },
      {
        "index": 1,
        "value": "B",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "B",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "1",
        "value": 1,
      },
    ],
  },
  {
    "childId": "C",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 0,
    "keys": [],
    "label": "C",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 2,
        "value": "C",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "C",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "2",
        "value": 2,
      },
    ],
  },
  {
    "childId": "A",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 1,
    "keys": [],
    "label": "A",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 2,
        "value": "C",
      },
      {
        "index": 0,
        "value": "A",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "A",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "1",
        "value": 1,
      },
    ],
  },
  {
    "childId": "B",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 1,
    "keys": [],
    "label": "B",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 2,
        "value": "C",
      },
      {
        "index": 1,
        "value": "B",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "B",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "1",
        "value": 1,
      },
    ],
  },
]
`;

exports[`Retain hierarchy even with arbitrary names getLegendItems special case: one input, one label 1`] = `
[
  {
    "childId": "A",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 0,
    "keys": [],
    "label": "A",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 0,
        "value": "A",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "A",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "1",
        "value": 1,
      },
    ],
  },
  {
    "childId": "A",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 1,
    "keys": [],
    "label": "A",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 0,
        "value": "A",
      },
      {
        "index": 0,
        "value": "A",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "A",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "1",
        "value": 1,
      },
    ],
  },
]
`;

exports[`Retain hierarchy even with arbitrary names getLegendItems special case: one input, two labels 1`] = `
[
  {
    "childId": "C",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 0,
    "keys": [],
    "label": "C",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 0,
        "value": "C",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "C",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "1",
        "value": 1,
      },
    ],
  },
  {
    "childId": "B",
    "color": "rgba(128, 0, 0, 0.5)",
    "depth": 1,
    "keys": [],
    "label": "B",
    "path": [
      {
        "index": 0,
        "value": "__null_small_multiples_key__",
      },
      {
        "index": 0,
        "value": "__root_key__",
      },
      {
        "index": 0,
        "value": "C",
      },
      {
        "index": 0,
        "value": "B",
      },
    ],
    "seriesIdentifiers": [
      {
        "key": "B",
        "specId": "spec1",
      },
    ],
    "values": [
      {
        "label": "1",
        "value": 1,
      },
    ],
  },
]
`;
